packages/go_router/example/lib/stateful_nested_navigation.dart

```diff
@@ -0,0 +1,420 @@
+              },
+              child: const Text('Push b'),
+            ),
+            const Padding(padding: EdgeInsets.all(4)),
+            if (secondDetailsPath != null)
+              TextButton(
+                onPressed: () {
+                  GoRouter.of(context).go(secondDetailsPath!);
+                },
+                child: const Text('View more details'),
+              ),
+          ],
+        ),
+      ),
+    );
+  }
+}
+
+/// The details screen for either the A or B screen.
+class DetailsScreen extends StatefulWidget {
+  /// Constructs a [DetailsScreen].
+  const DetailsScreen({
+    required this.label,
+    this.param,
+    this.withScaffold = true,
+    Key? key,
+  }) : super(key: key);
+
+  /// The label to display in the center of the screen.
+  final String label;
+
+  /// Optional param
+  final String? param;
+
+  /// Wrap in scaffold
+  final bool withScaffold;
+
+  @override
+  State<StatefulWidget> createState() => DetailsScreenState();
+}
+
+/// The state for DetailsScreen
+class DetailsScreenState extends State<DetailsScreen> {
+  int _counter = 0;
+
+  @override
+  Widget build(BuildContext context) {
+    if (widget.withScaffold) {
+      return Scaffold(
+        appBar: AppBar(
+          title: Text('Details Screen - ${widget.label}'),
+        ),
+        body: _build(context),
+      );
+    } else {
+      return _build(context);
+    }
+  }
+
+  Widget _build(BuildContext context) {
+    return Center(
+      child: Column(
+        mainAxisSize: MainAxisSize.min,
+        children: <Widget>[
+          if (widget.param != null)
+            Text('Parameter: ${widget.param!}',
+                style: Theme.of(context).textTheme.titleLarge),
+          const Padding(padding: EdgeInsets.all(4)),
+          Text('Details for ${widget.label} - Counter: $_counter',
+              style: Theme.of(context).textTheme.titleLarge),
+          const Padding(padding: EdgeInsets.all(4)),
+          TextButton(
+            onPressed: () {
+              setState(() {
+                _counter++;
+              });
+            },
+            child: const Text('Increment counter'),
+          ),
+          if (!widget.withScaffold) ...<Widget>[
+            const Padding(padding: EdgeInsets.all(16)),
+            TextButton(
+              onPressed: () {
+                GoRouter.of(context).pop();
+              },
+              child: const Text('< Back',
+                  style: TextStyle(fontWeight: FontWeight.bold, fontSize: 18)),
+            ),
+          ]
+        ],
+      ),
+    );
+  }
+}
+
+/// Builds a nested shell using a [TabBar] and [TabBarView].
+class TabbedRootScreen extends StatelessWidget {
+  /// Constructs a TabbedRootScreen
+  const TabbedRootScreen({Key? key}) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    final StatefulShellRouteState shellState = StatefulShellRoute.of(context);
+    final int branchCount = shellState.branchState.length;
```

Comment:
It would probably be more realistic to access `ShellRouteBranchState`'s routeBranch or rootRoutePath here rather than just using the index.

Reply:
Did you have something like this in mind? 

```dart
    final StatefulShellRouteState shellState = StatefulShellRoute.of(context);
    final List<TabbedRootScreenTab> children = shellState.branchState
        .mapIndexed((int index, ShellRouteBranchState e) =>
            TabbedRootScreenTab(branchState: e, index: index))
        .toList();
```

Reply:
Actually, that won't work, due to a "quirk" in the way TabBarView works. The created TabbedRootScreenTab widgets need to fetch the state fresh in the build method, via the index. 

But I see now that you perhaps meant that we could make the TabbedRootScreenTab use a path to match against the routeBranches? We could do that, but right now TabbedRootScreen works in kind of a generic (and agnostic) way when it comes to the TabBarView children, in that is doesn't really care what they contain, since this is anyway defined elsewhere (in the route definitions). What we however could do is to make TabbedRootScreen take the names of the tabs as input, that way the tab names would be defined closer to the definitions of what the contain. An even more realistic solution would probably be to create a subclass of ShellRouteBranch (like TabBranch) which could contain the name of the tab. That way, TabbedRootScreen could be even more "dumber". But maybe this is just making the example unnecessarily complex/verbose.